File: A3.cpp
#include <iostream>
#include <time.h>
using namespace std;

struct Pair
{
    int min;
    int max;
};

struct Pair getMinMax(int arr[], int n)
{
    struct Pair minmax;
    int i;

    if (n == 1)
    {
        minmax.max = arr[0];
        minmax.min = arr[0];
        return minmax;
    }

    if (arr[0] > arr[1])
    {
        minmax.max = arr[0];
        minmax.min = arr[1];
    }
    else
    {
        minmax.max = arr[1];
        minmax.min = arr[0];
    }

    for (i = 2; i < n; i++)
    {
        if (arr[i] > minmax.max)
            minmax.max = arr[i];

        else if (arr[i] < minmax.min)
            minmax.min = arr[i];
    }
    return minmax;
}

int main()
{
    clock_t tStart = clock();
    int i, len;
    cout << "Enter the Array length: ";
    cin >> len;
    int arr[len];
    cout << "\n";
    cout << "Enter the Number for array :";
    for (i = 0; i < len; i++)
    {
        cin >> arr[i];
    }
    int arr_size = len;

    struct Pair minmax = getMinMax(arr, arr_size);

    cout << "Minimum element is "
         << minmax.min << endl;
    cout << "Maximum element is "
         << minmax.max;

    printf("\nTime taken: %.9fs\n", (double)(clock() - tStart) / CLOCKS_PER_SEC);
    return 0;
}

File: A3.java
class A3 {
    public int max, min;

    public A3(int max, int min) {
        this.max = max;
        this.min = min;
    }

    public static void findMinAndMax(int[] A, int left, int right, A3 p) {

        if (left == right) {
            if (p.max < A[left]) {
                p.max = A[left];
            }

            if (p.min > A[right]) {
                p.min = A[right];
            }

            return;
        }

        if (right - left == 1) {
            if (A[left] < A[right]) {
                if (p.min > A[left]) {
                    p.min = A[left];
                }

                if (p.max < A[right]) {
                    p.max = A[right];
                }
            } else {
                if (p.min > A[right]) {
                    p.min = A[right];
                }

                if (p.max < A[left]) {
                    p.max = A[left];
                }
            }

            return;
        }

        int mid = (left + right) / 2;

        findMinAndMax(A, left, mid, p);

        findMinAndMax(A, mid + 1, right, p);
    }

    public static void main(String args[]) {
        int[] A = { 7, 2, 9, 3, 1, 6, 7, 8, 4 };

        A3 p = new A3(Integer.MIN_VALUE, Integer.MAX_VALUE);
        findMinAndMax(A, 0, A.length - 1, p);

        System.out.println("The minimum array element is " + p.min);
        System.out.println("The maximum array element is " + p.max);
    }
}